File: simulation/dtos/watchtower.py
from dataclasses import dataclass, field
from typing import Dict, List, Any, Optional

@dataclass
class IntegrityDTO:
    m2_leak: int
    fps: float

@dataclass
class MacroDTO:
    gdp: int
    cpi: float
    unemploy: float
    gini: float

@dataclass
class FinanceRatesDTO:
    base: float
    call: float
    loan: float
    savings: float

@dataclass
class FinanceSupplyDTO:
    m0: int
    m1: int
    m2: int
    velocity: float

@dataclass
class FinanceDTO:
    rates: FinanceRatesDTO
    supply: FinanceSupplyDTO

@dataclass
class PoliticsApprovalDTO:
    total: float
    low: float
    mid: float
    high: float

@dataclass
class PoliticsStatusDTO:
    ruling_party: str
    cohesion: float

@dataclass
class PoliticsFiscalDTO:
    revenue: int
    welfare: int
    debt: int

@dataclass
class PoliticsDTO:
    approval: PoliticsApprovalDTO
    status: PoliticsStatusDTO
    fiscal: PoliticsFiscalDTO

@dataclass
class PopulationDistributionDTO:
    q1: float
    q2: float
    q3: float
    q4: float
    q5: float

@dataclass
class PopulationMetricsDTO:
    birth: float
    death: float

@dataclass
class PopulationDTO:
    distribution: PopulationDistributionDTO
    active_count: int
    metrics: PopulationMetricsDTO

@dataclass
class WatchtowerSnapshotDTO:
    tick: int
    timestamp: float
    status: str
    integrity: IntegrityDTO
    macro: MacroDTO
    finance: FinanceDTO
    politics: PoliticsDTO
    population: PopulationDTO

File: frontend/src/types/watchtower.ts
export interface IntegrityDTO {
    m2_leak: number;
    fps: number;
}

export interface MacroDTO {
    gdp: number;
    cpi: number;
    unemploy: number;
    gini: number;
}

export interface FinanceRatesDTO {
    base: number;
    call: number;
    loan: number;
    savings: number;
}

export interface FinanceSupplyDTO {
    m0: number;
    m1: number;
    m2: number;
    velocity: number;
}

export interface FinanceDTO {
    rates: FinanceRatesDTO;
    supply: FinanceSupplyDTO;
}

export interface PoliticsApprovalDTO {
    total: number;
    low: number;
    mid: number;
    high: number;
}

export interface PoliticsStatusDTO {
    ruling_party: string;
    cohesion: number;
}

export interface PoliticsFiscalDTO {
    revenue: number;
    welfare: number;
    debt: number;
}

export interface PoliticsDTO {
    approval: PoliticsApprovalDTO;
    status: PoliticsStatusDTO;
    fiscal: PoliticsFiscalDTO;
}

export interface PopulationDistributionDTO {
    q1: number;
    q2: number;
    q3: number;
    q4: number;
    q5: number;
}

export interface PopulationMetricsDTO {
    birth: number;
    death: number;
}

export interface PopulationDTO {
    distribution: PopulationDistributionDTO;
    active_count: number;
    metrics: PopulationMetricsDTO;
}

export interface WatchtowerSnapshotDTO {
    tick: number;
    timestamp: number;
    status: string;
    integrity: IntegrityDTO;
    macro: MacroDTO;
    finance: FinanceDTO;
    politics: PoliticsDTO;
    population: PopulationDTO;
}
